Contributing Documentation - closes #121 (#159)
bob-crutchley/notes: CONTRIBUTING.md

# Contributing Guidelines
Anyone is welcome to collaborate on this project, but please adhere to the conventions in place.
<!--TOC_START-->
### Contents
- [Concepts](#concepts)
- [Naming Conventions](#naming-conventions)
- [Topic Structure](#topic-structure)
- [Modules](#modules)
- [Review Process](#review-process)
- [Weekly Process](#weekly-process)
	- [Roles](#roles)
		- [Courseware Admin](#courseware-admin)
		- [Project Admin](#project-admin)
	- [Breakdown](#breakdown)
		- [Monday (am)](#monday-am)
		- [Tuesday (pm)](#tuesday-pm)
		- [Wednesday (pm)](#wednesday-pm)
		- [Thursday (pm)](#thursday-pm)
		- [Friday (pm)](#friday-pm)
- [Other Implementation Guidelines](#other-implementation-guidelines)
	- [One Line Per Sentence](#one-line-per-sentence)

<!--TOC_END-->
## Concepts
All courseware on this project is in small and digestable modules.
These modules are then organised into topics.
Going forward Courses and Pathways will be introduced.
- Topics

>     This is a subject, which could be a programming language or tool. For example, Java, Jenkins, Kubernetes and Google Cloud Platform would all be considered subjects.

Reviewer: ```suggestion
    - This is a subject, which could be a programming language or tool. For example, Java, Jenkins, Kubernetes and Google Cloud Platform would all be considered subjects.
```

- Modules
    - Modules are the topics broken down into much smaller, digestible content and tasks.
- Courses
    - A course is a carefully ordered set of modules, from any topics, that are relevant for the course outcome.
- Pathways
    - Pathways are courses that have been put in order.
## Naming Conventions
Files and folders should be in lowercase with exceptions to the `README.md` file.
Names with multiple words should be seperated with a `-`.
## Topic Structure
Any example files & projects can be stored in the module folder.
There shouldn't be any reason for binary files to be on this repository, other than images at this point.

> If binary files are generated from examples, then be sure to include them in the `.gitignore`.  Images should not be uploaded, otherwise this project will be a very big download.  You can, of course, still use images, but source them from somewhere else on the internet (such as a public storage bucket).

Reviewer: You need to include instruction on how to do this at some point - im sure we all could, but i want the easiest cleanest way to maintain, what happens if we host the images on a personal account and then leave?

Author: I'll make a separate issue for this.

Author: Made an issue here: #160

The main `README.md` for a topic must at least have a heading (`# [TOPIC_NAME]`) and an overview section (`## Overview`):
```markdown
# Topic Name
## Overview
Some simple explanation or statement about the topic here.
```
A topic must have at least 1 module, usually this would be an introduction module to get started.
## Modules
A module should be a very small deliverable piece of information with a relevant task.
The only required headings are `## Overview` and `## Tasks`.
You must include an `## Overview` section because of the automation in place.
The `## Tasks` heading must be included because every module to should have a related task to complete.

Some modules will be very difficult to include a task because there might not be an easy way to implement what has been discussed in the module.
To get around this quizzes, Google Form questionnaires and links to videos can be included as a task.

## Review Process
Modules will go through 2 sets of reviewing before being released into the main branch.

The first review is for technical and implementation purposes, therefore they would be reviewed ideally by someone with some knowledge in the area:
- Do the tasks work and are clear?

> - Are the statements and explained concepts correct throughout the module

Reviewer: ```suggestion
- Are the statements and explained concepts correct throughout the module?
```

> - Does it serve its purpose well, as a piece of material for learning

Reviewer: ```suggestion
- Does it serve its purpose well, as a piece of material for learning?
```

- Does it make sense grammatically, are there any spelling errors etc.

The second set of reviewing is more for grammatical purposes and general looks of the material:

> - Any spelling mistakes

Reviewer: ```suggestion
- Any spelling mistakes.
```

> - Grammatical improvements to ensure it makes sense and for readability

Reviewer: ```suggestion
- Grammatical improvements to ensure it makes sense and for readability.
```

> - Formatting improvements to improve readability, this could be things like heading sizes and line breaks

Reviewer: ```suggestion
- Formatting improvements to improve readability, this could be things like heading sizes and line breaks.
```

## Weekly Process
Internally a weekly process is being formed for this project which allows modules to be implemented on a weekly basis.
This process only really concerns internal contributors of the project.

### Roles
#### Courseware Admin
This person reviews pull requests from release candidates into the master branch.
The courseware admin has the main say in what issues and modules are focused on each week and has the most visibility of the available courseware on the project.
#### Project Admin
Project admins are responsible for managing the automation and overall direction for the project.

> 

Reviewer: Do we need more info about the other 'developer' roles?

Author: I'll add something about internal developers

### Breakdown
Throughout the week pull requests are reviewed and merged to the `dev` branch, after Wednesday no more pull requests will be merged.
#### Monday (am)

> A project admin, courseware admin and available internal contributors have a meeting to discuss and establisht he following:

Reviewer: ```suggestion
A project admin, courseware admin and available internal contributors have a meeting to discuss and establish he following:
```

> A project admin, courseware admin and available internal contributors have a meeting to discuss and establisht he following:

Reviewer: ```suggestion
A project admin, courseware admin and available internal contributors have a meeting to discuss and establish the following:
```

- What upcoming training requirements are there for material
- What modules need to be created
- Prioritise the modules and issues that need to be implemented
- Created a milestone for the next version, typically just incrementing the revision
- Decide who (internally) is responsible for completing those modules by **Wednesday**
#### Tuesday (pm)
- Stand-up meeting to check internal progress
#### Wednesday (pm)
- Cut off for merging to the `dev` branch.
- Any modules not reviewed and merged into the `dev` branch are taken from the release.
- Release candidate branch and a pull request to the `master` branch is created with a name of `R-[VERSION]`, `R-0.0.3` for instance.
#### Thursday (pm)
- Stand-up meeting to check internal progress on release candidate reviews.
#### Friday (pm)
- Revert any incomplete reviews and merge the release candidate branch to `master`
- Rebase the `dev` branch with the release candidate branch
- Delete the release branch.
- Tag the master branch as the new version.

> ## Other Implementation Guidelines

Reviewer: A note here perhaps about implemeting multiple modules as part of one PR as you mentioned previously

Author: done

		- [Courseware Developers & Maintainers (Internal)](#courseware-developers--maintainers-internal)
	- [Breakdown](#breakdown)
		- [Monday (am)](#monday-am)
		- [Tuesday (pm)](#tuesday-pm)
		- [Wednesday (pm)](#wednesday-pm)
		- [Thursday (pm)](#thursday-pm)
		- [Friday (pm)](#friday-pm)
- [Other Implementation Guidelines](#other-implementation-guidelines)
	- [One Line Per Sentence](#one-line-per-sentence)
	- [Pull Requests and Issues for 1 Module](#pull-requests-and-issues-for-1-module)

<!--TOC_END-->
## Concepts
All courseware on this project is in small and digestable modules.
These modules are then organised into topics.
Going forward Courses and Pathways will be introduced.
- Topics
    - This is a subject, which could be a programming language or tool. For example, Java, Jenkins, Kubernetes and Google Cloud Platform would all be considered subjects.
- Modules
    - Modules are the topics broken down into much smaller, digestible content and tasks.
- Courses
    - A course is a carefully ordered set of modules, from any topics, that are relevant for the course outcome.
- Pathways
    - Pathways are courses that have been put in order.
## Naming Conventions

> Files and folders should be in lowercase with exceptions to the `README.md` file.

Reviewer 2: ```suggestion
Files, folders and packages should be in lowercase with exceptions to the `README.md` file.
```

> Names with multiple words should be seperated with a `-`.

Reviewer 2: ```suggestion
Names with multiple words should be separated with a `-` e.g cloud-enabling-technologies
```

## Topic Structure
Any example files & projects can be stored in the module folder.
There shouldn't be any reason for binary files to be on this repository, other than images at this point.

> If binary files are generated from examples, then be sure to include them in the `.gitignore`.  Images should not be uploaded, otherwise this project will be a very big download.  You can, of course, still use images, but source them from somewhere else on the internet (such as a public storage bucket).

Reviewer 2: Mention imgur as a place to host images easily

Here is an example layout of a topic:
```text
./topics/example-topic/
├── modules
│   ├── module-1
│   │   └── README.md - Include broken-down concepts and tasks here
│   └── module-2
│       └── README.md
└── README.md - Overview of the topic, you could maybe include installation guides here if they aren't to cumbersome
```
The main `README.md` for a topic must at least have a heading (`# [TOPIC_NAME]`) and an overview section (`## Overview`):
```markdown
# Topic Name
## Overview
Some simple explanation or statement about the topic here.
```
A topic must have at least 1 module, usually this would be an introduction module to get started.
## Modules
A module should be a very small deliverable piece of information with a relevant task.

> The only required headings are `## Overview` and `## Tasks`.

Reviewer 2: I would like to see module name as a mandatory heading. 

# Module name

## Overview

### Something to learn

## Tasks

Author: done

You must include an `## Overview` section because of the automation in place.
The `## Tasks` heading must be included because every module to should have a related task to complete.

Some modules will be very difficult to include a task because there might not be an easy way to implement what has been discussed in the module.

> To get around this quizzes, Google Form questionnaires and links to videos can be included as a task.

Reviewer 2: Could also mention hidden answers

Author: done

## Review Process
Modules will go through 2 sets of reviewing before being released into the main branch.

The first review is for technical and implementation purposes, therefore they would be reviewed ideally by someone with some knowledge in the area:
- Do the tasks work and are clear?
- Are the statements and explained concepts correct throughout the module?
- Does it serve its purpose well, as a piece of material for learning?
- Does it make sense grammatically, are there any spelling errors etc.

> 

Reviewer 2: Are abbreviations explained?

Author: I've added this to #161

Reviewer 2: resolve this one then pal

#### Monday (am)
A project admin, courseware admin and available internal contributors have a meeting to discuss and establish he following:
- What upcoming training requirements are there for material
- What modules need to be created
- Prioritise the modules and issues that need to be implemented
- Created a milestone for the next version, typically just incrementing the revision
- Decide who (internally) is responsible for completing those modules by **Wednesday**
#### Tuesday (pm)
- Stand-up meeting to check internal progress
#### Wednesday (pm)
- Cut off for merging to the `dev` branch.
- Any modules not reviewed and merged into the `dev` branch are taken from the release.
- Release candidate branch and a pull request to the `master` branch is created with a name of `R-[VERSION]`, `R-0.0.3` for instance.
#### Thursday (pm)
- Stand-up meeting to check internal progress on release candidate reviews.
#### Friday (pm)
- Revert any incomplete reviews and merge the release candidate branch to `master`
- Rebase the `dev` branch with the release candidate branch
- Delete the release branch.
- Tag the master branch as the new version.

> ## Other Implementation Guidelines

Reviewer 2: Add a developer guide

Author: made an issue for this #161
